File: bots/intgrah/v54.7.1/builder/hooks/propagate_symmetry.py
from __future__ import annotations

from typing import TYPE_CHECKING

from cambc import Environment
from marker import MarkerSymmetry
from util.constants import MAX_WIDTH
from util.directions import DIR8

if TYPE_CHECKING:
    from cambc import Controller

    from builder import Builder


def end_of_turn_propagate_symmetry(self: Builder, ct: Controller) -> None:
    """If we've collapsed to a single symmetry, drop a marker on a
    nearby tile with no existing building so other units that see it
    converge too. Markers are the lowest-priority building — we'd never
    destroy anything to place one, so we only place on tiles that are
    already unbuilt."""
    if self.symmetry is None:
        return
    payload = MarkerSymmetry(symmetry=self.symmetry).encode()
    for d in DIR8:
        target = self.my_pos.add(d)
        if not self.in_bounds(target):
            continue
        i = target.y * MAX_WIDTH + target.x
        if self.env[i] == Environment.WALL:
            continue
        if self.buildings[i] is not None:
            continue
        if ct.can_place_marker(target):
            ct.place_marker(target, payload)
            return

File: bots/rs/v55/src/builder/hooks/propagate_symmetry.rs
//! Translation of `bots/intgrah/v54.7.9/builder/hooks/propagate_symmetry.py`.

use cambc::{Controller, ControllerApi, Environment};

use crate::builder::Builder;
use crate::marker::Marker;
use crate::util::constants::MAX_WIDTH;
use crate::util::directions::DIR8;

/// If we've collapsed to a single symmetry, drop a marker on a
/// nearby tile with no existing building so other units that see it
/// converge too. Markers are the lowest-priority building — we'd never
/// destroy anything to place one, so we only place on tiles that are
/// already unbuilt.
pub fn end_of_turn_propagate_symmetry(builder: &mut Builder, ct: &mut Controller<'_>) {
    let Some(symmetry) = builder.symmetry() else {
        return;
    };
    let payload = Marker::Symmetry { symmetry }.encode();
    for &d in &DIR8 {
        let target = builder.state.my_pos.add(d);
        if !builder.in_bounds(target) {
            continue;
        }
        let i = (target.y as usize) * MAX_WIDTH + (target.x as usize);
        if builder.env[i] == Some(Environment::Wall) {
            continue;
        }
        if builder.buildings[i].is_some() {
            continue;
        }
        if ct.can_place_marker(target).unwrap() {
            ct.place_marker(target, payload).unwrap();
            return;
        }
    }
}
